src/program.cpp

```diff
@@ -89,7 +89,16 @@ void program::assign(const program& p)
     }
     impl->ctx         = p.impl->ctx;
     impl->target_name = p.impl->target_name;
-    impl->modules     = p.impl->modules;
```

Comment:
Why is this line removed?

Reply:
We cannot directly copy the modules one by one. The design here considers all modules are of the same level, which dies not consider the lexical scoping issue. Actually, these modules are parent-child relationship in the ONNX model. 

By assigning the modules one after another, it may cause the issue that one submodule needs an input from its parent module, but its parent module is not copies yet, so that input is unavailable.
 
To avoid this problem, we have to start from the main module, then go to its child submodule, and so on..... In this way, when a submodule needs an input from its parent, we can ensure it is available.

Reply:
Looks like we are using a linear data structure to implement something like a tree, implementation of the functions like to_value/from_value/assign are a little weird.

Reply:
The module class should be copyable, but of course, the program class needs to update the references to the new modules.

I forgot about the instruction_refs that need to be updated. We can add a `replace_refs` method to the instruction class that takes two maps for module_ref and instruction_ref. This would update any reference(in inputs(), outputs(), and module_args) in the instruction class to use the new reference. We can then copy the program with something like this:

```cpp
void program::assign(const program &p)
{
  if (!impl)
    impl = std::make_unique<program_impl>();
  impl->ctx = p.impl->ctx;
  impl->target_name = p.impl->target_name;
  // Copy the modules using its copy contructor
  impl->modules = p.impl->modules;
  // Build a map from old module to new module
  std::unordered_map<module_ref, module_ref> mod_map;
  std::transform(
      impl->modules.begin(), impl->modules.end(), p.impl->modules.begin(),
      std::back_inserter(mod_map),
      [](auto &&x, auto &&y) { return std::make_pair(&x.second, &y.second); });
  // Build a map from old ins to new ins
  std::unordered_map<instruction_ref, instruction_ref> ins_map;
  for (auto &&pp : mod_map)
  {
    auto old_ins = iterator_for(pp.first);
    auto new_ins = iterator_for(pp.second);
    std::transform(old_ins.begin(), old_ins.end(), new_ins.begin(),
                   std::back_inserter(ins_map),
                   [](auto x, auto y) { return std::make_pair(x, y); });
  }
  // Update all references from all modules
  for (auto &&mp : impl->modules)
  {
    for (auto ins : iterator_for(mp.second))
      ins->replace_refs(mod_map, ins_map);
  }
}
```

I think from_value/to_value should be removed from the module class since it doesn't make sense for the module class to serialize its data when it references data from other classes. Instead,

- Modules are stored as an array instead of an object
- The modules are sorted into a breadth-first order so the parent module is always read or written first.
- The name of the inputs and outputs should be prefixed with the module name(ie `main:@1` instead of just `@1`).

> Looks like we are using a linear data structure to implement something like a tree

I am not sure if a tree would help. First, a tree implies all the nodes are connected which may not be the case with dead branches. Secondly, a tree could manage the relationships between modules but it wouldn't manage relationships between instructions.

This is what makes it more complicated. If we just needed relationships between modules we could use module names instead(as the name would be stable and wouldn't cause a dangling reference). However, if we want lexical scoping, then we need to update instruction_refs as well.